File: MaxFlow.java
import java.util.LinkedList;

class MaxFlow {
    static final int V = 5; // Número de vértices en el grafo

    // Retorna verdadero si hay un camino desde el nodo fuente 's'
    // hasta el nodo sumidero 't' en el grafo residual. También llena
    // el arreglo 'parent[]' para almacenar el camino.
    boolean bfs(int rGraph[][], int s, int t, int parent[]) {
        boolean visited[] = new boolean[V];
        for (int i = 0; i < V; ++i) {
            visited[i] = false;
        }

        LinkedList<Integer> queue = new LinkedList<Integer>();
        queue.add(s);
        visited[s] = true;
        parent[s] = -1;

        while (queue.size() != 0) {
            int u = queue.poll();

            for (int v = 0; v < V; v++) {
                if (!visited[v] && rGraph[u][v] > 0) {
                    if (v == t) {
                        parent[v] = u;
                        return true;
                    }
                    queue.add(v);
                    parent[v] = u;
                    visited[v] = true;
                }
            }
        }
        return false;
    }

    // Retorna el flujo máximo desde el nodo fuente 's' hasta el nodo
    // sumidero 't' en el grafo dado.
    int fordFulkerson(int graph[][], int s, int t) {
        int u, v;

        int rGraph[][] = new int[V][V];
        for (u = 0; u < V; u++) {
            for (v = 0; v < V; v++) {
                rGraph[u][v] = graph[u][v];
            }
        }

        int parent[] = new int[V];
        int max_flow = 0;

        while (bfs(rGraph, s, t, parent)) {
            int path_flow = Integer.MAX_VALUE;
            for (v = t; v != s; v = parent[v]) {
                u = parent[v];
                path_flow = Math.min(path_flow, rGraph[u][v]);
            }
            for (v = t; v != s; v = parent[v]) {
                u = parent[v];
                rGraph[u][v] -= path_flow;
                rGraph[v][u] += path_flow;
            }
            max_flow += path_flow;
        }
        return max_flow;
    }

    // Función principal para probar el algoritmo de flujo máximo
    public static void main(String[] args) {
        int graph[][] = new int[][] {
            {0, 20, 30, 10, 0},
            {0, 0, 40, 0, 30},
            {0, 0, 0, 10, 20},
            {0, 0, 5, 0,20},
            {0, 0, 0, 0, 0},
        };
        MaxFlow m = new MaxFlow();
        System.out.println("El flujo máximo posible es " + m.fordFulkerson(graph, 0, 4));
    }
}

File: MaxFlow.py
from collections import defaultdict

class Graph:
    def __init__(self, graph):
        self.graph = graph # grafo residual
        self.ROW = len(graph)

    def BFS(self, s, t, parent):
        visited = [False]*(self.ROW)
        queue = []
        queue.append(s)
        visited[s] = True

        while queue:
            u = queue.pop(0)
            for ind, val in enumerate(self.graph[u]):
                if visited[ind] == False and val > 0:
                    queue.append(ind)
                    visited[ind] = True
                    parent[ind] = u
                    if ind == t:
                        return True
        return False
    
    def FordFulkerson(self, source, sink):
        parent = [-1]*(self.ROW)
        max_flow = 0
        while self.BFS(source, sink, parent) :
            path_flow = float("Inf")
            s = sink
            while(s != source):
                path_flow = min (path_flow, self.graph[parent[s]][s])
                s = parent[s]
            max_flow += path_flow
            v = sink
            while(v != source):
                u = parent[v]
                self.graph[u][v] -= path_flow
                self.graph[v][u] += path_flow
                v = parent[v]
        return max_flow

# Solicitamos al usuario ingresar el número de nodos
n = int(input("Ingrese el número de nodos: "))

# Creamos una matriz de adyacencia inicialmente vacía
graph = []
for i in range(n):
    # Solicitamos al usuario que ingrese los valores para cada fila de la matriz
    row = list(map(int, input(f"Ingrese los valores para la fila {i+1}, separados por espacios: ").split()))
    graph.append(row)

# Creamos el grafo a partir de la matriz de adyacencia
g = Graph(graph)

# Solicitamos al usuario que ingrese el nodo fuente y el nodo sumidero
source = int(input("Ingrese el nodo fuente: "))
sink = int(input("Ingrese el nodo sumidero: "))

# Mostramos el resultado
print ("El flujo máximo posible es %d " % g.FordFulkerson(source, sink))
